### DISCUSSION TITLE
TFA QR code data should be improved

### DISCUSSION THREAD

**Discussion by wendevlin** [directus/directus#6129] [Feature Requests]
Hello,

When I was working on my TFA PR yesterday I saw that the QR code for adding TFA to an Authenticator app is generated like this:

Firstname + Lastname of the user, the application name hardcoded ¨Directus"

`api/src/services/authentication.ts`
```
async generateOTPAuthURL(pk: string, secret: string): Promise<string> {
  const user = await this.knex.select('first_name', 'last_name').from('directus_users').where({ id: pk }).first();
  const name = `${user.first_name} ${user.last_name}`;
  return authenticator.keyuri(name, 'Directus', secret);
}
```

I see 2 problems here:
1. A default admin user doesn´t have first and lastname set
2. If you have multiple directus projects it is always called directus.

Here is a screenshot of authy app adding a user without first and lastname set:

![image](https://user-images.githubusercontent.com/12148533/121152837-d82c1b80-c845-11eb-99e1-e4836c217a34.png)

Of course the user can always change it but it would be more convenient to have it already set unique.

my suggestion would be:

**Projectname + user-email** or **Directus - Projectname: user-email**

**Comment by wendevlin** [comment]
@rijkvanzanten if it's fine I would add this change of the QR data to my open PR https://github.com/directus/directus/pull/6117 tomorrow.

**Reply by rijkvanzanten** [reply]
Yeah, sounds good!

**Reply by wendevlin** [reply]
I updated the method, but I found out that the automatic icon generation just works if you write `Directus`. When you add the project name it doesn't work.

So I just changed firstname + lastname to user.email. The user is still able to add the project name later on if she/he has multiple two factor logins at different directus instances.

#6117

**Reply by rijkvanzanten** [reply]
> but I found out that the automatic icon generation just works if you write Directus.

What do you mean? I'm a little confused what you mean by "automatic icon generation"

**Reply by wendevlin** [reply]
Sorry. 
When you have a look in the screenshot above you see the directus rabbit. This is the Authy app, a popular Authenticator app.

I actually don't know where Authy is getting the directus icon, I just found a docu about how to change it in within the app.

But when I used a title like "Directus - My project" the icon became a screnshot of a login error.

Moreover the otplib is referencing to this documentation from the Google Authenicator: https://github.com/google/google-authenticator/wiki/Key-Uri-Format
The info here about the issuer: "STRONGLY RECOMMENDED: The issuer parameter is a string value indicating the provider or service this account is associated with"

So when we want to follow this step, I would say directus is the service that is used, not the project name.

**Reply by rijkvanzanten** [reply]
Ahh I see. I'd rather not rely on whatever Authy decides is the right thing to do for the logo tbh. As a matter of fact, this icon is the logo from years ago, which is pretty outdated.

As for the name, I agree that "Directus" is generally the service, but on the same thought, Directus should be as white-label as possible, so I think we should go with the project name after all. Could you tell the issuer where to find the logo? We could point them to the same logo used on the login page!

**Reply by wendevlin** [reply]
Yes you are right, we shouln't rely on Authy.
The issuer is just a string, you cannot define a logo or anything else.

Logos are searched by the Authenticator apps itself. Authy seems to use google image search. (When I search for directus, I see the current logo as first result, but who knows their search parameters)

So I would keep it as you have changed it.

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
